### PR TITLE
Testa/fs stream

### PR DESCRIPTION
**Pull request by testaco**
for issue #53

### CODE REVIEW

**Comment by aschulm** [comment]
Blocked by #61

File: fw/drivers/sd.c [hunk 1]
```diff
@@ -284,15 +307,123 @@ int sd_write_block_start(void* block, uint32_t block_num) //{{{
...
+        // contains the token, but I doubt this happens
+        if (rx != 0x00)
+        {
+            halt(ERROR_SD_CMD25_ACK_FAILED);
+        }
+
+        multi_write_in_progress = 1;
+    }
+
+	// write data token
```
Comment (testaco) [inline fw/drivers/sd.c]: Fixed.

File: fw/drivers/sd.c [hunk 2]
```diff
@@ -284,15 +307,123 @@ int sd_write_block_start(void* block, uint32_t block_num) //{{{
...
+        tx = 0xFD;
+        usart_spi_txrx(&USARTE1, &tx, NULL, 1);
+
+        // let the sd card have some time to raise the busy flag
+        usart_spi_txrx(&USARTE1, NULL, &rx, 1);
+
+        // read until the busy flag is cleared 
+        rx = 0;
+        tries = 0;
+        while (rx == 0 && tries < SD_MAX_BUSY_TRIES)
```
Comment (testaco) [inline fw/drivers/sd.c]: I upped it since on cold start the self test was failing with this timing out.

File: fw/drivers/sd.c [hunk 3]
```diff
@@ -301,9 +432,18 @@ int sd_write_block_update() //{{{
...
 
-		 //give the SD card 8 clock cycles to prepare for command
-		usart_spi_txrx(&USARTE1, NULL, NULL, 1);
-		gpio_on(&PORTE, SPI_SS_PIN_bm);
+#ifdef DEBUG_SD_TIMING
+        printf("write_block: %d\n", down_ms);
+#endif
+
+
+        // Toggle slave select if this is a single block write
```
Comment (testaco) [inline fw/drivers/sd.c]: Fixed.

File: fw/drivers/sd.c [hunk 4]
```diff
@@ -284,15 +307,123 @@ int sd_write_block_start(void* block, uint32_t block_num) //{{{
 	return 1;
 } //}}}
 
+int sd_write_multi_block_start(void* block, int block_num) {
```
Comment (testaco) [inline fw/drivers/sd.c]: Fixed.

File: fw/drivers/sd.c [hunk 5]
```diff
@@ -313,14 +472,14 @@ int sd_write_block_update() //{{{
...
-	int i = 0;
+	int i = 0, j = 0;
 	uint8_t block[SD_BLOCK_LEN];
 	uint32_t* block_int = (uint32_t*)block;
 
 	printf("sd self test\n");
 
 	printf("the inserted SD card's capacity is %lu blocks\n",
-	       sd_capacity());
+		   sd_capacity());
```
Comment (aschulm) [inline fw/drivers/sd.c]: extra spaces?

File: fw/drivers/sd.c [hunk 6]
```diff
@@ -343,5 +502,45 @@ int sd_self_test() //{{{
...
+	for (j = 0; j < TEST_MULTI_BLOCK_COUNT; ++j)
+	{
+		memset(block, 0, sizeof(block));
+		sd_read_block(block, 250 + j);
+		for (i = 0; i < (SD_BLOCK_LEN >> 2); i++)
+		{
+			if (block_int[i] != i)
+			{
+				printf("FAILED wrote[%d] read[%lu]\n",
+						i,
```
Comment (aschulm) [inline fw/drivers/sd.c]: extra tab?

File: fw/fs.c [hunk 7]
```diff
@@ -316,14 +316,18 @@ int fs_close() //{{{
 
 	// there is one more partial block that remains to be written
 	uint32_t last_block_idx = file_startblock_idx + BYTES_TO_BLOCKS(file_byte_idx);
```
Comment (aschulm) [inline fw/fs.c]: this should be moved inside of the if statement

**Pull request merged**